components/eam/src/physics/crm/pam/pam_driver.cpp

```diff
@@ -14,26 +13,90 @@
-  using yakl::atomicAdd;
   using yakl::c::parallel_for;
   using yakl::c::SimpleBounds;
+  using yakl::atomicMax;
+  //------------------------------------------------------------------------------------------------
+  auto nens          = coupler.get_option<int>("ncrms");
+  auto gcm_nlev      = coupler.get_option<int>("gcm_nlev");
+  auto crm_nz        = coupler.get_option<int>("crm_nz");
+  auto crm_nx        = coupler.get_option<int>("crm_nx");
+  auto crm_ny        = coupler.get_option<int>("crm_ny");
+  auto crm_dx        = coupler.get_option<real>("crm_dx");
+  auto crm_dy        = coupler.get_option<real>("crm_dy");
+  auto &dm_device = coupler.get_data_manager_device_readonly();
+  auto &dm_host   = coupler.get_data_manager_host_readonly();
+  auto uvel       = dm_device.get<real const,4>("uvel");
+  auto wvel       = dm_device.get<real const,4>("wvel");
+  auto input_zint = dm_host.get<real const,2>("input_zint").createDeviceCopy();
+  //------------------------------------------------------------------------------------------------
+  yakl::ParallelMax<real,yakl::memDevice> pmax( crm_nz*nens );
+  real cfl = 0;
+  int num_subcycle = 1;
+  int constexpr max_num_subcycle = 10;
+  real2d wvel_max("wvel_max",crm_nz,nens);
+  real2d uvel_max("uvel_max",crm_nz,nens);
+  real2d cfl_max("cfl_max",  crm_nz,nens);
+  //------------------------------------------------------------------------------------------------
+  // initialize max U and W arrays
+  parallel_for( SimpleBounds<2>(crm_nz,nens) , YAKL_LAMBDA (int k, int n) {
+    wvel_max(k,n) = 0.0;
+    uvel_max(k,n) = 0.0;
+  });
+  // calculate max U and W
+  parallel_for( SimpleBounds<4>(crm_nz,crm_ny,crm_nx,nens) , YAKL_LAMBDA (int k, int j, int i, int n) {
+    yakl::atomicMax(uvel_max(k,n), sqrt(uvel(k,j,i,n)*uvel(k,j,i,n)) );
+    yakl::atomicMax(wvel_max(k,n), fabs(wvel(k,j,i,n)) );
+  });
+  // find max CFL between horizontal and vertical CFL values
+  parallel_for( SimpleBounds<2>(crm_nz,nens) , YAKL_LAMBDA (int k, int n) {
+    int k_gcm = gcm_nlev-1-k;
+    real crm_dz = input_zint(k_gcm,n) - input_zint(k_gcm+1,n);
+    real cfl_u = uvel_max(k,n)*crm_dt_fixed/crm_dx;
+    real cfl_w = wvel_max(k,n)*crm_dt_fixed/crm_dz;
+    cfl_max(k,n) = max(cfl_u,cfl_w);
+  });
+  // calculate final CFL across ensemble
+  real cfl_loc = pmax(cfl_max.data());
+  cfl = max(cfl,cfl_loc);
+  // update number of subcycles and time step
+  num_subcycle = max(num_subcycle,max(1,static_cast<int>(ceil(cfl/0.7))));
+  real crm_dt_subcycle = crm_dt_fixed / num_subcycle;
+  coupler.set_option<real>("crm_dt",crm_dt_subcycle);
+  // check for excessive subcylcing - don't exit, just print
+  if(num_subcycle > max_num_subcycle) {
+    real umax = pmax(uvel_max.data());
+    real wmax = pmax(wvel_max.data());
+    printf("PAM_DRIVER - WARNING: excessive subcycling!"
+          " - num_subcycle: %3.3d  dt: %8.4f  cfl: %8.4f  umax: %8.2f  wmax: %8.2f \n",
+          num_subcycle,crm_dt_subcycle,cfl,umax,wmax);
```

Comment:
You can `std::cout << std::endl` at the end to force the buffer to flush to file / terminal. I like to do that at the end of all warning or error messages to make it much more likely it gets output before the model dies.

Reply:
Cool, thanks for the tip!